### PR TITLE
Add Loss Averaging and no_backward_sync() in Gradient Accumulation Logic

### PR DESCRIPTION
**Pull request by waitzkin**
This fixes the gradient accumulation part in finetune directory.

I think averaging the loss is needed, based on the definition of gradient accumulation. 

Also, for efficiency in fine-tuning, I added no_backward_sync() where it wasn't.

I would appreciate any comments about this :)

### CODE REVIEW

File: finetune/lora.py [hunk 1]
```diff
@@ -108,9 +108,10 @@ def train(
         t0 = time.time()
 
         input_ids, targets = get_batch(fabric, train_data)
-        logits = model(input_ids)
-        loss = loss_fn(logits, targets)
-        fabric.backward(loss)
+        with fabric.no_backward_sync(model, enabled=((iter_num + 1) % gradient_accumulation_iters != 0)):
+            logits = model(input_ids)
+            loss = loss_fn(logits, targets)
+            fabric.backward(loss / gradient_accumulation_iters )
```
Comment (carmocca) [inline finetune/lora.py:114]:
```suggestion
            fabric.backward(loss / gradient_accumulation_iters)
```

**Review by carmocca — APPROVED** [review]
Great catch!

cc @awaelchli

**Review by awaelchli — APPROVED** [review]
Thanks. I remember having fixed moving the model forward under the ctx manager, but somehow the wrong version has multiplied itself into other scripts anyway 😄

File: finetune/lora.py [hunk 2]
```diff
@@ -108,9 +108,10 @@ def train(
         t0 = time.time()
 
         input_ids, targets = get_batch(fabric, train_data)
-        logits = model(input_ids)
-        loss = loss_fn(logits, targets)
-        fabric.backward(loss)
+        with fabric.no_backward_sync(model, enabled=((iter_num + 1) % gradient_accumulation_iters != 0)):
```
Comment (awaelchli) [inline finetune/lora.py:111]: FWY, I didn't add it here because this script was meant to be running on a single GPU. Anyway, doesn't hurt to have it.
